python/ray/experimental/signal.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import hashlib
+import time
+
+import ray
+
+SYSTEM_ID_STRING = ray.ray_constants.ID_SIZE * b"\xaa"
+SYSTEM_ID = ray.ObjectID(SYSTEM_ID_STRING)
+
+SIG_USER = 100
+
+class Signal(object):
+    """Signal object"""
+    def __init__(self, type, value):
+        self.sig_type = type
+        self.sig_value = value
+    def value(self):
+        return self.sig_value
+    def type(self):
+        return self.sig_type
+
+def _get_signal_object_id_from_binary(source_id, counter):
+    m = hashlib.sha1()
+    m.update(source_id)
+    m.update(bytes(counter))
+    return ray.ObjectID(m.digest())
```

Comment:
@stephanie-wang, thanks for the comments. 

Regarding (2) we also need GCS because by default we want to get the old signals. Also, I think that we should optimize the first version for simplicity. Two more questions:
1. Is there a simple way to get the task id? Right now I have a function `_get_tesk_id_from_object_id(object_id)` to do it.
2. Do you have an example of calling `ComputeReturnId` or `ComputePutId` from Python?

Reply:
Also, a few things about fault tolerance:
1. The typical use case of is for the raylet/bakend to send a signal (i.e., SIG_FAULT) on the behalf of an actor/task that has failed, i.e., `signal.send(signal.Signal(SIG_FAULT), source_id = a)`. In this case we might want to actually disable fault tolerance for the actors/tasks.
2. If there is a user level signal, and if the actor/task sending it fails and the signal is lost, I think we are probably ok, as the backend will send a SIG_FAULT on behalf of the failed actor/task. So I don't think we need to reconstruct the lost signal. However, we do need to ignore it which the current code doesn't do.

Reply:
Can't you get the old signals by just iterating through the signals until one of them times out? Then, I don't think the GCS is necessary.

1. You should replace that method with the C++ extension for Python. It actually looks like that particular method is already implemented [here](https://github.com/ray-project/ray/blob/master/src/ray/raylet/lib/python/common_extension.cc#L815).
2. I don't think we have an example of this in the code right now, but it will be similar to the above example for getting the task ID from the object ID.

For fault tolerance: Right I agree, but that actually wouldn't be possible with the object IDs generated out-of-band, since the backend cannot associate the signal object ID with the actor/task that failed.